Switch to sslcrypto for cryptography tasks

purplesyringa opened the pull request:
[sslcrypto](https://github.com/imachug/sslcrypto) is a new MIT-licensed library I've recently made. It supports AES and ECC. I originally made it to get rid of GPLv3 pyelliptic, but interestingly it turned out even more useful.

First, it replaces pyelliptic, pybitcointools and a part of CryptMessage so there's less code involved. Second, it supports both OpenSSL backend and pure-Python fallback both for AES and ECC (pyelliptic doesn't support pure-Python ECC so CryptMessage didn't work on some machines). Third, it's licensed under MIT (I can add any other license you choose) so we wouldn't have problems if we wanted to re-license a part of ZeroNet code under some lax license. Fourth, it supports *all* OpenSSL versions since 0.9.8 (other cryptography libraries support a lot less). Fifth, its fallback implementation is a bit safer than pybitcointools because it fixes a vulnerability (I don't want to name it here to avoid possible private key leaks).

Sixth, it's fast. Plain performance tests say that it's about as fast as pyelliptic (both for AES and ECC) and that its fallback implementation is 1.5 times faster than pybitcointools (ECIES only, ECDSA is about the same). However, some tests I ran on ZeroMail show that message decryption is about 2 or 3 times faster than the current version (I've only run tests on single-thread model because more threads are slower for me for some reason both for sslcrypto and your version).

Additionally, sslcrypto is [tested](https://github.com/imachug/sslcrypto/actions) on Python 3.4 to 3.8 with OpenSSL 0.9.8, 1.0.0, 1.0.1, 1.0.2, 1.1.0 and 1.1.1.

/cc @filips123

---

This PR closes #1974, closes #917 and closes #993.

purplesyringa committed:
Use sslcrypto instead of pyelliptic and pybitcointools

purplesyringa committed:
Fix CryptMessage

purplesyringa committed:
Support Python 3.4

purplesyringa committed:
Fix user creation

purplesyringa committed:
Get rid of pyelliptic and pybitcointools

purplesyringa committed:
Fix typo

purplesyringa committed:
Delete test file

purplesyringa commented:
Ignore the failing checks, it's just a coincidence (@HelloZeroNet I guess you'll have to increase some constants)

ghost commented:
Don't include this **until the license is not fixed!** 
Be warned! I going to release publicly ZeroNet under **GPLv3+** immediately after if this get merged!

filips123 commented:
> Don't include this until the license is not fixed!

Why?

ghost commented:
Because I have a copy of ZeroNet where these libraries are not replaced and **I will not replace them! I will be forced to release ZeroNet under GPLv3+ immediately!** 
@HelloZeroNet do not merge this until the licensing issue not solved! Also note that most code is not originates from @imachug he copied most of the code from other people including from pastebin! 

The voting was a big show to get some time to write up this library.

filips123 commented:
> Also note that most code is not originates from @imachug he copied most of the code from other people including from pastebin!

Prove this.

ghost commented:
I already proven that he is copied from others and so this library he claims his own is nothing more but a **compilation** and can't be licenses under MIT either!

filips123 commented:
1. You can't include MIT license in MIT license? That's something new...

2. So you think it is not possible to write cryptography-related library because somebody already wrote it and now whole cryptography operations are copyrighted?

ghost commented:
> now whole cryptography operations are copyrighted?

**Not only copyrighted but also patented!**

filips123 commented:
And you own that patents, right?

Why do we already have more than one cryptography library?

ghost commented:
> And you own that patents, right?

No but I can use patented works as long I don't claim it to be mine! 

> Why do we already have more than one cryptography library?

Why now before the license change need to change it? This still not solve the licensing issue! And even if all libraries replaced **I can publish my version of ZeroNet under GPLv3+ immediately!** 

**You are all risking here loosing all your copyrights in the previously released ZeroNet versions!**

ghost commented:
@HelloZeroNet 
**Do not merge this until the licensing issue is not resolved or you seriously damage the ZeroNet project!** Merging this will give immediately permission for all forks to use either GPLv3+ or GPLv3-only on ZeroNet!

filips123 commented:
> No but I can use patented works as long I don't claim it to be mine!

Where does @imachug claims that all code is his? Where some code was taken from other people, this is clearly stated in that file.

And still, I don't know what is the difference with other cryptography libraries...

> This still not solve the licensing issue!

Yes it does. If incompatible libraries are replaced, incompatibilities are fixed.

> I can publish my version of ZeroNet under GPLv3+ immediately!

No you can't. You need agreement from contributors for that.

> Merging this will give immediately permission for all forks to use either GPLv3+ or GPLv3-only on ZeroNet!

WTF are you talking?

ghost commented:
Replacing the library in order to avoid changing the license **will not replace the library in my copy and other 1,900 forks!** 

Under GPL I have already granted all copyright permission to modify and distribute the software under GPL but legally I can't distribute ZeroNet under GPLv2! I'm guaranteed the right to distribute the software with or without my modifications and this right is not revocable! 

This means if you replace the library, I'm getting a green light to release ZeroNet under a **compatible** GPL license of my choosing, which is GPLv3+!

ghost commented:
I don't need a shit from contributors! I already granted all rights under GPL! 
**You need to tell contributors that you are fooled them!** You made up a shady, misleading and illegal voting to fool them and gain some time to write a library! 

I urge you @HelloZeroNet to don't even think about to merge this until the license is not fixed or you get into serious trouble!

ghost commented:
@imachug 
> Third, it's licensed under MIT (I can add any other license you choose) so we wouldn't have problems if we wanted to re-license a part of ZeroNet code under some lax license.

You can't re-license **part** of ZeroNet. Under [GPL section 2](https://github.com/HelloZeroNet/ZeroNet/issues/2273#issuecomment-560026695) the **entire ZeroNet** is already released under GPL! 
You also can't claim ownership over your contributions to a work which is licensed under GPL! 
There is no such thing as relicensing your lines what you added to the source code!

ghost commented:
> ..pybitcointools because it fixes a vulnerability (I don't want to name it here to avoid possible private key leaks).

You are a joke! Now you telling me that `pybitcointools` is vulnerable to private key leaks? I want to see what they will say in BitcoinTalk about this! :rofl: **[Theymos](https://bitcointalk.org/index.php?action=profile;u=35)** will be happy to hear it!

Send Theymos and email and tell him what you found! [theymos@mm.st](mailto:theymos@mm.st)

ghost commented:
![ballmer_peak](https://user-images.githubusercontent.com/58347673/70005239-c1c24700-1560-11ea-8cf2-530a76f8c382.png)

purplesyringa commented:
> Replacing the library in order to avoid changing the license will not replace the library in my copy and other 1,900 forks!

Other forks would be illegal but at least the official one would be fine.

purplesyringa commented:
> You can't re-license part of ZeroNet

I can't, but nofish can relicense some of his code. Not as if it's related to this PR... 

> Theymos will be happy to hear it!

Who's that guy? AFAIK all/most pybitcointools forks are abandoned.

filips123 commented:
> Replacing the library in order to avoid changing the license will not replace the library in my copy and other 1,900 forks!

Changing the license to GPLv3+ would magically replace it in forks?

> Under GPL I have already granted all copyright permission to modify and distribute the software under GPL but legally I can't distribute ZeroNet under GPLv2!

You have right to distribute it under GPLv2, not any GPL license. That's why it includes GPLv2 license file, not just GPL license file.

> This means if you replace the library, I'm getting a green light to release ZeroNet under a compatible GPL license of my choosing, which is GPLv3+!

Replacing the library means that (almost) all incompatibilities are fixed and if you want to comply with licenses, you need to download new libraries.

> I don't need a shit from contributors! I already granted all rights under GPL!

As FSF said, contributions only granted distribution under GPLv2.

> You need to tell contributors that you are fooled them!

Because we allowed them to do what they should be able to do?

> Under GPL section 2 the entire ZeroNet is already released under GPL!

No it is not. It is licensed under GNU General Public License **version 2**.

> You also can't claim ownership over your contributions to a work which is licensed under GPL!

You "care" so much about contributions and then you would ignore their opinion and agreement when changing the license...

ghost commented:
@filips123 I never seen more retarded person in my life than you! I hope @imachug at least have some brain and can translate this to you. 

- **NOBODY CAN DISTRIBUTE ZERONET USING GPLv2, THAT WOULD BE ILLEGAL AND CRIMINAL!** 
- **CONTRIBUTORS ALREADY GRANTED ALL RIGHTS TO KEEP DISTRIBUTING ZERONET UNDER GPL!** 
- **THIS RIGHTS TO SHARE AND MODIFY ARE NOT REVOCABLE!** 

Because, I do have **irrevocable** rights to share the code and I can't do it legally under GPLv2, I'm obligated by law to share ZeroNet under **GPLv3+**! 

Contributors agreed to release the **entire** ZeroNet under GPL! Under section two of the GPL the **entire** ZeroNet is under the same license, contributors can't change the license on they portion of the code and certainly they can't claim proprietary rights!

ghost commented:
> Who's that guy? AFAIK all/most pybitcointools forks are abandoned.

**Copyright is protected for 70 years!**

filips123 commented:
Why do we already have other libraries than pybitcointools?

HelloZeroNet commented:
Thanks, looking great. I will do some benchmark on it. To add new dependency we need to improve the updater script that check if the user has installed the required libs, notify is something is missing and also need to find a solution for macOS where the runtime is not updated right now.
As a temporary solution we can include this package to src/lib.

Please limit the license conversation to the related GitHub issue. (I will change the license once I see the license vote/active line numbers report)

purplesyringa commented:
Cool! Please share benchmark results later. As for the libraries, can we somewhat simulate pip by downloading new packages to, say, src/lib/_pip and using them if pip is not available?

filips123 commented:
Simulating Pip properly is quite hard. The best solution we can currently do is to just place your library into `lib` manually (like already done for some dependencies and like I did for DNS plugin).

For more proper solution, we need to switch distributing system for ZeroNet quite a bit (check my existing issue about this).

purplesyringa committed:
Add sslcrypto to tree

purplesyringa commented:
Hey @HelloZeroNet, I've moved sslcrypto to src/lib. Please take a look at it now.

filips123 commented:
@imachug Even if it is in `src/lib`, I think that it is still useful to have is in `requirements.txt` because new installation will use it (@HelloZeroNet ZeroBundle also needs to be updated) and after some time, it could be probably removed from repository.

purplesyringa commented:
I think I will update sslcrypto soon (refactoring, you know) so I'd better not have two possible versions (one from PyPI and one from the bundle). I'd remove sslcrypto from requirements.txt for now and add it when we get better packaging working.

purplesyringa committed:
Update sslcrypto

purplesyringa committed:
Add pyaes to src/lib

purplesyringa commented:
I've just uploaded a refactored sslcrypto version and also added pyaes to as a dependency.

filips123 commented:
@HelloZeroNet Any update?

HelloZeroNet commented:
Currently I'm working on a multi-thread related modification. (it should be done in the next days) 

I will test this PR after it's done.

HelloZeroNet commented:
Thanks, it's looking great and fast, my results on Windows:

```

OpenSSL 1.0.2s  28 May 2019 (currently bundled)

Old (2.94s):
* Running testCryptEciesEncrypt x 100 .......... Done in 0.716s = Fine (1.68x)
* Running testCryptEciesDecrypt x 500 .......... Done in 1.044s = Fine (1.25x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.707s = OK (0.96x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.276s = OK (0.98x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.201s = Fine (1.25x)

New (1.96s):
* Running testCryptEciesEncrypt x 100 .......... Done in 0.387s = WOW (3.10x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.695s = Fast (1.87x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.500s = Fine (1.36x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.223s = Fine (1.21x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.162s = Fine (1.55x)


OpenSSL 1.1.0j  20 Nov 2018 (bundled with Python 3.7)

Old (6.63s):
* Running testCryptEciesEncrypt x 100 .......... Done in 1.698s = Goodish (0.71x)
* Running testCryptEciesDecrypt x 500 .......... Done in 2.502s = Ehh (0.52x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 1.921s = Ehh (0.35x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.300s = OK (0.90x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.214s = OK (1.17x)

New (3.73s):
* Running testCryptEciesEncrypt x 100 .......... Done in 0.888s = Fine (1.35x)
* Running testCryptEciesDecrypt x 500 .......... Done in 1.432s = OK (0.91x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.942s = Goodish (0.72x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.264s = OK (1.02x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.213s = OK (1.17x)


OpenSSL 1.1.1d  10 Sep 2019 (bundled with Python 3.8)

Old (2.55s):
* Running testCryptEciesEncrypt x 100 .......... Done in 0.610s = Fast (1.97x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.894s = Fine (1.45x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.558s = Fine (1.22x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.285s = OK (0.95x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.209s = OK (1.20x)

New (1.99s):
* Running testCryptEciesEncrypt x 100 .......... Done in 0.419s = WOW (2.86x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.683s = Fast (1.90x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.452s = Fine (1.50x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.241s = OK (1.12x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.204s = Fine (1.22x)  
```

So it's 25-100% faster, than the current lib we use.

As for improvement the result of sslcrypto.ecc.get_backend() could include the path of the loaded openssl dll/so file.

HelloZeroNet commented on plugins/CryptMessage/CryptMessagePlugin.py:
```diff
@@ -200,8 +187,8 @@ def testCryptEciesDecrypt(self, num_run=1):
             assert len(aes_key) == 32
-            ecc = CryptMessage.getEcc(self.privatekey)
-            assert ecc.decrypt(encrypted) == self.utf8_text.encode("utf8"), "%s != %s" % (ecc.decrypt(encrypted), self.utf8_text.encode("utf8"))
+            decrypted = CryptMessage.eciesDecrypt(base64.b64encode(encrypted))
```
missing privatekey argument, it should be: `decrypted = CryptMessage.eciesDecrypt(base64.b64encode(encrypted), self.privatekey)`

purplesyringa replied:
There, fixed.

purplesyringa committed:
Fix typo in tests

purplesyringa commented:
@HelloZeroNet Can you merge this now?

HelloZeroNet commented:
Yeah, I will test it on ARM and macOS, then merge it.

HelloZeroNet commented:
Hm looks like in multi-threaded crypt mode it segfaults on Linux. First discovered on ARM, but on x86 it's also does.

```
ZeroNet-sslcrypto$ zeronet.py --threads_crypt 2 test benchmark --filter cryptEcies
- Starting ZeroNet...
- Libsecpk256k1 loaded: CompiledLib in 0.024s
- sslcrypto loaded in 0.049s
- Version: 0.7.1 r4322, Python 3.6.8 (default, Oct  7 2019, 12:59:55)
[GCC 8.3.0], Gevent: 1.4.0
- Running benchmark
* Running testCryptEciesEncrypt x 100 .......... Done in 0.186s = Insane!! (6.45x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.337s = Insane!! (3.86x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) Segmentation fault (core dumped)

ZeroNet-py3$ python3 zeronet.py --threads_crypt 2 test benchmark --filter cryptEcies
- Starting ZeroNet...
- Libsecpk256k1 loaded: CompiledLib in 0.023s
- Version: 0.7.1 r4325, Python 3.6.8 (default, Oct  7 2019, 12:59:55)
[GCC 8.3.0], Gevent: 1.4.0
- Running benchmark
* Running testCryptEciesEncrypt x 100 .......... Done in 0.341s = Insane!! (3.52x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.487s = WOW (2.67x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.297s = Fast (2.29x)
* Result:
 - Total: 3 tests
 - Success: 3 tests
 - Failed: 0 tests
 - Average speed factor: 2.83x (WOW)
* Test benchmark done in 1.190s
```

(It would be usefull to print the openssl version and the loaded lib path on the sslcrypt loaded line)

Update: I was also able to reproduce the problem on Windows with `py -3 zeronet.py --threads_crypt 2 test benchmark --filter cryptEciesDecryptMulti --num_multipler 10`

purplesyringa commented:
O_o I'll check that, thanks.

purplesyringa committed:
Update sslcrypto version

purplesyringa commented:
@HelloZeroNet I've added OpenSSL path to `get_backend` and also made sslcrypto thread-safe. Please check it again.

purplesyringa commented:
O_o It still segfaults on Python 3.4, wtf.

purplesyringa commented:
The tests were restarted for some reason so they might pass, but please don't merge the PR before I figure out the problem!

HelloZeroNet commented:
Python 3.4 support is not essential, so we can limit the crypto threads to 1 there if you can't figure out the solution. I will also test it later today

purplesyringa commented:
Hm, I'm not sure whether it was a Python 3.4 bug or it was a global highly-unlikely-to-reproduce bug. I'd better stress-test this a bit more.

HelloZeroNet commented:
For some reason on Python 3.4 sslcurve.private_to_address(priv, is_compressed=False) returns string instead of bin which makes https://github.com/HelloZeroNet/ZeroNet/blob/6c013493c21eb40deeb34a844295b61f2bd0b454/src/Crypt/CryptBitcoin.py#L62 fail.
Also would be better to use privatekey_bin variable name for decoded private key to make it more clear that it's not the argument that passed.

HelloZeroNet commented:
Thanks, looks like the segfault got fixed!

purplesyringa commented:
> Thanks, looks like the segfault got fixed!

Not really, I haven't touched anything. I'd run it under valgrind but I don't have a linux machine right now.

HelloZeroNet commented:
I think that segfault was not connected with the library as (if i remember right) it happened in one of the site downloading tests where openssl library is not used at all

purplesyringa commented:
> For some reason on Python 3.4 sslcurve.private_to_address(priv, is_compressed=False) returns string instead of bin

Hm, are you sure? I've run sslcrypto and ZeroNet tests under Python 3.4.10 and I also tried to run the code under 3.4.0 but it always returns bytes. Do you have reproduction steps or something?

purplesyringa committed:
Use privatekey_bin instead of privatekey for bytes objects

purplesyringa commented:
> I think that segfault was not connected with the library as (if i remember right) it happened in one of the site downloading tests where openssl library is not used at all

I hope so.

purplesyringa commented:
@HelloZeroNet Ready to be merged.

purplesyringa commented:
I've just received [the old log](https://gist.github.com/ssabrii/495e4ff94b0cf39e16a831bc0805a498) from Travis CI support. It looks like Crypt *is* used in that test.

HelloZeroNet commented:
I will test it later today on ARM/macOS. You can ignore the download test fails, currently I'm working on refactoring the worker code, hopefully it will fix the issue.

HelloZeroNet commented:
It still failes on Python3.4, because return sslcurve.private_to_address(privatekey, is_compressed=False).decode()  drops "AttributeError: 'str' object has no attribute 'decode' " error.

It happens because for some reason on Py3.4:
```
>>> base58.b58encode_check(b"\x00")
'1Wh4bh'
```
On Py3.7:
```
>>> base58.b58encode_check(b"\x00")
b'1Wh4bh'
```

Update: nvm it was because of outdated base58 python package. It was changed in version 1.0.0: https://github.com/keis/base58/commit/871dbe14fe8e9675ef262a5e06fbf3874562a28c#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR4

HelloZeroNet commented:
What changed in the signature method that makes it generates different signs?
```
$ python3 zeronet.py test benchmark --filter testSign
! Error: AssertionError: HFGXaDauZ8vX/N9Jn+MRiGm9h+I94zUhDnNYFaqMGuOi+4+BbWHjuwmx0EaKNV1G+kP0tQDxWu0YApxwxZbSmZU= != G1GXaDauZ8vX/N9Jn+MRiGm9h+I94zUhDnNYFaqMGuOiBHB+kp4cRPZOL7l1yqK5BHa6J+W97bMjvTXtxzljp6w= in Benchmark/BenchmarkPlugin.py line 165 > Benchmark/BenchmarkPlugin.py line 316
```

HelloZeroNet commented:
Previously I have not tested the verification speed, as libsecp256k1 is the main lib we use there, but looks like it got affected in negative way:
```
ZeroNet-sslcrypto $ py -3 zeronet.py test benchmark --filter testVerify
* Running testVerify ({'lib_verify': 'sslcrypto_fallback'}) x 20 .......... Done in 0.059s = Insane!! (6.46x)
* Running testVerify ({'lib_verify': 'sslcrypto'}) x 200 .......... Done in 0.537s = Ehh (0.56x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.321s = Ehh (0.31x)
```

Using the current codebase:

```
ZeroNet-py3$ py -3 zeronet.py test benchmark --filter testVerify
* Running testVerify ({'lib_verify': 'btctools'}) x 20 .......... Done in 0.211s = Fast (1.80x)
* Running testVerify ({'lib_verify': 'openssl'}) x 200 .......... Done in 0.156s = Fast (1.93x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.050s = Fast (2.00x)
```

(you need to update the lib_verify names here: https://github.com/imachug/ZeroNet/blob/sslcrypto/plugins/Benchmark/BenchmarkPlugin.py#L106-L107)

purplesyringa committed:
Fix sslcrypto

purplesyringa committed:
Fix Benchmark plugin

purplesyringa committed:
Don't calculate the same thing twice

purplesyringa committed:
Only import sslcrypto once

purplesyringa committed:
Handle fallback sslcrypto implementation during tests

purplesyringa commented:
> What changed in the signature method that makes it generates different signs?

Fixed, thanks for reporting!

> Previously I have not tested the verification speed, as libsecp256k1 is the main lib we use there, but looks like it got affected in negative way:

The main reason for that is some legacy code I forgot to remove which made verification run twice. It's fixed now (and I've added a few more optimizations just in case).

Here are my results:

With sslcrypto:

```
$ py -3 zeronet.py test benchmark --filter testVerify
* Running testVerify ({'lib_verify': 'sslcrypto_fallback'}) x 20 .......... Done in 0.172s = Fast (2.21x)
* Running testVerify ({'lib_verify': 'sslcrypto'}) x 200 .......... Done in 1.295s = Sloooow (0.23x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.328s = Ehh (0.31x)
```

Without sslcrypto:

```
$ py -3 zeronet.py test benchmark --filter testVerify
* Running testVerify ({'lib_verify': 'btctools'}) x 20 .......... Done in 1.342s = Sloooow (0.28x)
* Running testVerify ({'lib_verify': 'openssl'}) x 200 .......... Done in 1.248s = Sloooow (0.24x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.328s = Ehh (0.31x)
```

So libsecp256k1 and OpenSSL are about as fast -- sometimes sslcrypto is faster than openssl, sometimes it's not but I think such slight differences don't matter much. The fallback implementation is several times faster.

HelloZeroNet commented:
Thanks, I can confirm the signature and the speed got fixed.
Win10:
```
ZeroNet-sslcrypto $ py -3 zeronet.py test benchmark
* Running testHdPrivatekey x 50 .......... Done in 0.041s = Insane!! (13.94x)
* Running testSign x 20 .......... Done in 0.021s = Insane!! (21.97x)
* Running testVerify ({'lib_verify': 'sslcrypto_fallback'}) x 20 .......... Done in 0.019s = Insane!! (20.02x)
* Running testVerify ({'lib_verify': 'sslcrypto'}) x 200 .......... Done in 0.151s = Fast (1.99x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.046s = Fast (2.18x)
* Running testPackMsgpack x 100 x 100 x 5KB .......... Done in 0.250s = Fine (1.40x)
* Running testCryptEciesEncrypt x 100 .......... Done in 0.221s = Insane!! (5.42x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.408s = WOW (3.19x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.277s = Fast (2.45x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.115s = Fast (2.35x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.085s = WOW (2.95x)
```
vs current codebase:
```
ZeroNet-py3 $ py -3 zeronet.py test benchmark
 Running testHdPrivatekey x 50 .......... Done in 0.364s = Fine (1.57x)
* Running testSign x 20 .......... Done in 0.274s = Fine (1.68x)
* Running testVerify ({'lib_verify': 'btctools'}) x 20 .......... Done in 0.204s = Fast (1.87x)
* Running testVerify ({'lib_verify': 'openssl'}) x 200 .......... Done in 0.151s = Fast (1.99x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.045s = Fast (2.23x)
* Running testCryptEciesEncrypt x 100 .......... Done in 0.488s = Fast (2.46x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.699s = Fast (1.86x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.424s = Fine (1.60x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.169s = Fine (1.60x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.133s = Fast (1.88x)
```

So for me libsecp256k1 still significantly faster, but for some reason the sslcrypto_fallback seems almost as fast as the non-fallback, so I suspect it's also using openssl is some form.

Will do some more tests tomorrow.

purplesyringa commented:
> So for me libsecp256k1 still significantly faster

This is actually correct. libsecp256k1 is a secp256k1-specific library which has lots of optimizations for this curve. On the other hand, OpenSSL supports all kinds of curves and doesn't have any curve-specific optimizations.

purplesyringa committed:
Fix sslcrypto fallback implementation selection

purplesyringa commented:
> but for some reason the sslcrypto_fallback seems almost as fast as the non-fallback, so I suspect it's also using openssl is some form

Damn, I'm really sorry for so many mistakes. It should be fixed now.

HelloZeroNet commented:
It also looks good on macOS:

```
* Running testVerify ({'lib_verify': 'sslcrypto_fallback'}) x 20 .......... Done in 0.091s = Insane!! (4.17x)
* Running testVerify ({'lib_verify': 'sslcrypto'}) x 200 .......... Done in 0.135s = Fast (2.22x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.044s = Fast (2.29x)
* Running testCryptEciesEncrypt x 100 .......... Done in 0.182s = Insane!! (6.59x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.359s = Insane!! (3.62x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.261s = WOW (2.60x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.210s = Fine (1.29x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.115s = Fast (2.18x)
```
vs current code base:
```
* Running testVerify ({'lib_verify': 'btctools'}) x 20 .......... Done in 0.196s = Fast (1.94x)
* Running testVerify ({'lib_verify': 'openssl'}) x 200 .......... Done in 0.165s = Fast (1.82x)
* Running testVerify ({'lib_verify': 'libsecp256k1'}) x 200 .......... Done in 0.046s = Fast (2.19x)
* Running testCryptEciesEncrypt x 100 .......... Done in 0.337s = Insane!! (3.56x)
* Running testCryptEciesDecrypt x 500 .......... Done in 0.484s = WOW (2.68x)
* Running testCryptEciesDecryptMulti x 5 x 100 (2 threads) ..... Done in 0.299s = Fast (2.27x)
* Running testCryptAesEncrypt x 10000 .......... Done in 0.286s = OK (0.94x)
* Running testCryptAesDecrypt x 10000 .......... Done in 0.173s = Fine (1.44x)
```

So everything looks good for me and I think we can merge it if you have nothing to add it yet.

btw yesterday there was a segfaulted test again 
on Python 3.4 on:
`plugins/Bigfile/Test/TestBigfile.py::TestBigfile::testNullFileRead[ipv6-3072] /home/travis/.travis/functions: line 109:  4464 Segmentation fault      (core dumped)`

[travis_log.txt](https://github.com/HelloZeroNet/ZeroNet/files/3963851/travis_log.txt)

It would be great to try reproduce the environment and see if we can trigger the segfault there.

purplesyringa commented:
> It would be great to try reproduce the environment and see if we can trigger the segfault there.

I don't have a Linux machine rn so I can't run valgrind. Can someone do that for me?

purplesyringa commented:
Yay thanks!

purplesyringa commented:
Hey @HelloZeroNet I managed to reproduce segfault on GitHub actions. Please revert this PR until I get rid of the problem.

filips123 commented:
How often are they reproducible? Because if they are rare, it might still not be needed to revert PR.

purplesyringa commented:
Here are the logs in case someone wants to debug that:

```
2019-12-15T16:42:34.8799674Z Fatal Python error: Segmentation fault
2019-12-15T16:42:34.8800244Z 
2019-12-15T16:42:34.8801375Z Current thread 0x00007fcb537fe700 (most recent call first):
2019-12-15T16:42:34.8806417Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 280 in new_private_key
2019-12-15T16:42:34.8811412Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 147 in new_private_key
2019-12-15T16:42:34.8827943Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 201 in encrypt
2019-12-15T16:42:34.8828534Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.8828966Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.8829457Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.8829887Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.8830315Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.8830591Z 
2019-12-15T16:42:34.8830890Z Thread 0x00007fcb53fff700 (most recent call first):
2019-12-15T16:42:34.8831286Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 53 in __init__
2019-12-15T16:42:34.8831701Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 252 in _point_to_affine
2019-12-15T16:42:34.8832117Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 296 in private_to_public
2019-12-15T16:42:34.8832529Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 151 in private_to_public
2019-12-15T16:42:34.8832936Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 213 in encrypt
2019-12-15T16:42:34.8833360Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.8835783Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.8840172Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.8845015Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.8849679Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.8850701Z 
2019-12-15T16:42:34.8852200Z Thread 0x00007fcb69a1c700 (most recent call first):
2019-12-15T16:42:34.8856400Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 313 in ecdh
2019-12-15T16:42:34.8915229Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 180 in derive
2019-12-15T16:42:34.8920763Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 294 in decrypt
2019-12-15T16:42:34.8981934Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.8984726Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.8988112Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.8993806Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.8998247Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.8998392Z 
2019-12-15T16:42:34.9000252Z Thread 0x00007fcb6b21f700 (most recent call first):
2019-12-15T16:42:34.9005176Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 294 in private_to_public
2019-12-15T16:42:34.9056521Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 151 in private_to_public
2019-12-15T16:42:34.9061130Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 204 in encrypt
2019-12-15T16:42:34.9065434Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.9069859Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.9074372Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.9083878Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.9141401Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.9142445Z 
2019-12-15T16:42:34.9143381Z Thread 0x00007fcb6921b700 (most recent call first):
2019-12-15T16:42:34.9144412Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 278 in new_private_key
2019-12-15T16:42:34.9145468Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 147 in new_private_key
2019-12-15T16:42:34.9146509Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 201 in encrypt
2019-12-15T16:42:34.9147534Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.9148557Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.9149855Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.9150932Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.9152003Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.9152907Z 
2019-12-15T16:42:34.9153814Z Thread 0x00007fcb6a21d700 (most recent call first):
2019-12-15T16:42:34.9154844Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 313 in ecdh
2019-12-15T16:42:34.9155869Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 180 in derive
2019-12-15T16:42:34.9156920Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 270 in decrypt
2019-12-15T16:42:34.9157396Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 106 in run
2019-12-15T16:42:34.9157782Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.9158167Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.9158597Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.9159508Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.9160077Z 
2019-12-15T16:42:34.9160393Z Thread 0x00007fcb6aa1e700 (most recent call first):
2019-12-15T16:42:34.9160918Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 278 in new_private_key
2019-12-15T16:42:34.9161391Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 147 in new_private_key
2019-12-15T16:42:34.9161792Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 201 in encrypt
2019-12-15T16:42:34.9162194Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.9162557Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.9162956Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.9163388Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.9164098Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.9165046Z 
2019-12-15T16:42:34.9165391Z Thread 0x00007fcb68a1a700 (most recent call first):
2019-12-15T16:42:34.9165809Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/openssl/ecc.py", line 294 in private_to_public
2019-12-15T16:42:34.9166215Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 151 in private_to_public
2019-12-15T16:42:34.9167229Z   File "/home/runner/work/sslcrypto/sslcrypto/sslcrypto/_ecc.py", line 213 in encrypt
2019-12-15T16:42:34.9167647Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 105 in run
2019-12-15T16:42:34.9168019Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 12 in run
2019-12-15T16:42:34.9168470Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 870 in run
2019-12-15T16:42:34.9169136Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 926 in _bootstrap_inner
2019-12-15T16:42:34.9169686Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 890 in _bootstrap
2019-12-15T16:42:34.9169964Z 
2019-12-15T16:42:34.9170268Z Thread 0x00007fcb72213700 (most recent call first):
2019-12-15T16:42:34.9170685Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 1060 in _wait_for_tstate_lock
2019-12-15T16:42:34.9171492Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/threading.py", line 1044 in join
2019-12-15T16:42:34.9173786Z   File "/home/runner/work/sslcrypto/sslcrypto/test/conf.py", line 22 in decorated
2019-12-15T16:42:34.9174607Z   File "/home/runner/work/sslcrypto/sslcrypto/test/test_thread_safety.py", line 109 in test_ecies
2019-12-15T16:42:34.9176172Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/python.py", line 166 in pytest_pyfunc_call
2019-12-15T16:42:34.9178191Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
2019-12-15T16:42:34.9179794Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
2019-12-15T16:42:34.9181548Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
2019-12-15T16:42:34.9183120Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
2019-12-15T16:42:34.9184591Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/python.py", line 1435 in runtest
2019-12-15T16:42:34.9186084Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 131 in pytest_runtest_call
2019-12-15T16:42:34.9187579Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
2019-12-15T16:42:34.9189564Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
2019-12-15T16:42:34.9192586Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
2019-12-15T16:42:34.9194936Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
2019-12-15T16:42:34.9197022Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 207 in <lambda>
2019-12-15T16:42:34.9199161Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 234 in from_call
2019-12-15T16:42:34.9201024Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 207 in call_runtest_hook
2019-12-15T16:42:34.9203338Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 182 in call_and_report
2019-12-15T16:42:34.9205151Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 96 in runtestprotocol
2019-12-15T16:42:34.9206937Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/runner.py", line 81 in pytest_runtest_protocol
2019-12-15T16:42:34.9208697Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
2019-12-15T16:42:34.9210723Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
2019-12-15T16:42:34.9212481Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
2019-12-15T16:42:34.9214247Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
2019-12-15T16:42:34.9216000Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/main.py", line 270 in pytest_runtestloop
2019-12-15T16:42:34.9217761Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
2019-12-15T16:42:34.9219861Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
2019-12-15T16:42:34.9221798Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
2019-12-15T16:42:34.9223531Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
2019-12-15T16:42:34.9225271Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/main.py", line 246 in _main
2019-12-15T16:42:34.9227401Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/main.py", line 196 in wrap_session
2019-12-15T16:42:34.9229545Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/main.py", line 239 in pytest_cmdline_main
2019-12-15T16:42:34.9231382Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
2019-12-15T16:42:34.9233130Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
2019-12-15T16:42:34.9234872Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
2019-12-15T16:42:34.9236592Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
2019-12-15T16:42:34.9238330Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/_pytest/config/__init__.py", line 92 in main
2019-12-15T16:42:34.9240429Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/pytest/__main__.py", line 7 in <module>
2019-12-15T16:42:34.9242202Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/coverage/execfile.py", line 229 in run
2019-12-15T16:42:34.9244152Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/coverage/cmdline.py", line 703 in do_run
2019-12-15T16:42:34.9246917Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/coverage/cmdline.py", line 555 in command_line
2019-12-15T16:42:34.9249586Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/lib/python3.7/site-packages/coverage/cmdline.py", line 824 in main
2019-12-15T16:42:34.9251736Z   File "/opt/hostedtoolcache/Python/3.7.5/x64/bin/coverage", line 8 in <module>
2019-12-15T16:42:41.2962217Z /home/runner/work/_temp/85e0b3f1-269e-45bc-8cb3-8af75c30d74a.sh: line 2:  9519 Segmentation fault      (core dumped) coverage run -m pytest test
2019-12-15T16:42:41.2962656Z test/test_thread_safety.py::test_ecies 
2019-12-15T16:42:41.2967753Z ##[error]Process completed with exit code 139.
2019-12-15T16:42:41.2997280Z Cleaning up orphan processes
```

purplesyringa commented:
Does anyone have a Linux machine they can share?

filips123 commented:
Can't you use virtual machine?

purplesyringa commented:
Now *that's* something I haven't thought about... I'll give it a chance.

HelloZeroNet commented:
I just sent you a VM ssh access in email

purplesyringa commented:
Thanks, I'll take a look soon. Please revert this PR though.

filips123 commented:
@HelloZeroNet @shortcutme @imachug Is there any update? did you figured out what caused segfault?

purplesyringa commented:
No idea.

HelloZeroNet commented:
It could be helpful to run tests with http://man.he.net/man1/catchsegv

purplesyringa commented:
I'm running stress tests on the VM... let's see if I get any results.

purplesyringa commented:
400 tests and... nothing. What am I expected to do now? valgrind doesn't detect any memory corruption as well.

HelloZeroNet commented:
You can add carchsegv to the tests + export the dumped core as artifact in case of segfault, then run the tests for a few time on travis/GitHub actions/circle and expect a segfault

filips123 commented:
@HelloZeroNet @imachug Any update here?
